### PR TITLE
Fix Diagnose invalid characters in bundle identifiers

### PR DESCRIPTION
**Pull request by agisilaos**
Fixes: #290

## Changes

This PR introduces bundle identifier validation in the `NavigatorIndex.Builder` class and adds corresponding unit tests to ensure its correctness.

### Implementation

- Added a private `validateBundleIdentifier()` method to `NavigatorIndex.Builder`.
- The method checks for invalid characters in the bundle identifier based on RFC 3986 URL host requirements.
- Invalid bundle identifiers trigger a warning diagnostic with a detailed explanation.

### Tests

- Created `DiagnosticBundleIdentifierTests` to thoroughly test the new validation logic.
- Tests cover valid bundle identifiers, identifiers with spaces, and identifiers with both valid and invalid special characters.
- Implemented using the GIVEN-WHEN-THEN structure for improved readability and maintainability.

## Rationale

1. **Improved Data Integrity**: Validating bundle identifiers ensures that only valid identifiers are used, preventing potential issues downstream in the documentation generation process.

2. **Compliance with Standards**: By adhering to RFC 3986 for URL hosts, we ensure that our bundle identifiers are compatible with web-based systems and avoid potential conflicts or encoding issues.

3. **Early Error Detection**: Catching invalid bundle identifiers early in the process allows developers to address issues promptly, improving the overall quality of documentation projects.

4. **Clear Feedback**: The detailed error messages provide developers with actionable information to correct invalid bundle identifiers.

5. **Robust Testing**: The comprehensive test suite ensures that the validation logic works as expected and helps prevent regressions in future updates.

## Impact
- Developers will receive immediate feedback if they use invalid characters in bundle identifiers.
- This change may catch existing invalid bundle identifiers in projects, requiring updates to conform to the new validation rules.
- The documentation generation process will be more robust against potential issues caused by malformed bundle identifiers.

## Notes for Reviewers
- Please pay special attention to the character set used for validation. Are there any additional characters we should consider allowing or restricting?
- The error message content and severity (warning) are open for discussion. Should we consider making this a more severe error?
- The test coverage aims to be comprehensive. Are there any additional edge cases you think we should test?

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary

### CODE REVIEW

File: Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift [hunk 1]
```diff
@@ -612,8 +612,40 @@ extension NavigatorIndex {
...
+        ///
+        /// - Note: Bundle identifiers must not contain spaces, and they must conform to the character set
+        ///         allowed for URL hosts as defined by RFC 3986.
+        ///
+        /// - SeeAlso: [RFC 3986](https://tools.ietf.org/html/rfc3986)
+        ///
+        /// - Important: This method assumes that `bundleIdentifier` is a non-optional string.
+        private func validateBundleIdentifier() {
+            // URL host valid characters (RFC 3986)
+            let validCharacters = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-._~"))
```
Comment (d-ronnqvist) [inline Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift:634]: It would be better to use `CharacterSet.urlHostAllowed` for this.

File: Tests/SwiftDocCTests/Diagnostics/DiagnosticBundleIdentifierTests.swift [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        let builder = NavigatorIndex.Builder(
+            outputURL: outputURL,
+            bundleIdentifier: "com.example.valid-bundle_identifier"
+        )
+
+        // WHEN
+        builder.setup()
+
+        // THEN
+        XCTAssertTrue(builder.problems.isEmpty, "No problems should be reported for a valid bundle identifier")
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Diagnostics/DiagnosticBundleIdentifierTests.swift:28]: The "given", "when", "then" comments is not a style that we use anywhere else. 

These specific tests won't be relevant—since the navigator index isn't the right place to verify the bundle identifier—but for any other tests, please omit those specific comments to match the style of the existing tests.

File: Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift
Comment (d-ronnqvist) [inline Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift:1]: The navigator index is just one of the places that make use of the bundle identifier but it's not a source of it. As such, it is not the right place to perform this validation.

**Review by d-ronnqvist — CHANGES_REQUESTED** [review]
Thank you for opening this PR. However, the `NavigatorIndex` is not the right place to validate the bundle identifier.

It would be better to validate the bundle identifier earlier, before the build starts.

There are only two places where the developer can specify a bundle identifier:
- A value in the documentation catalog's Into.plist
- A `--fallback-bundle-identifier` value on the command line

It would be a nicer experience for the developer if the diagnostic let the developer know which of these two that the invalid identifier came from.

[This code][c1] in [DocumentationBundle+Info.swift][f1] could be one place to consider adding this validation since it knows if the value came from the Info.plist or from the command line fallback.

[c1]: https://github.com/swiftlang/swift-docc/blob/main/Sources/SwiftDocC/Infrastructure/Workspace/DocumentationBundle%2BInfo.swift#L223-L238
[f1]: https://github.com/swiftlang/swift-docc/blob/main/Sources/SwiftDocC/Infrastructure/Workspace/DocumentationBundle%2BInfo.swift

When I tested running `docc convert /path/to/SomeCatalog.docc --fallback-bundle-identifier "not valid"` just now, it crashed failing to create a `ResolvedTopicReference` value. Because of this, a "warning" level diagnostic wouldn't be suitable, since that lets the build progress until it crashes.

That said, since we can diagnose the issue and know which characters are not allowed, it's possible that we could correct the issue by removing the disallowed characters in the developer-provided bundle identifier (or by replacing them with an allowed character like "-"). Doing so would avoid the crash which would allow for a "warning" level diagnostic again. 

----

It's also worth noting that the `docc process-archive index` command has its own bundle identifier command line argument that possibly would require its own validation.

**Comment by d-ronnqvist** [comment]
@agisilaos Are you still working on this? 
I'm touching some other code in this area and am wondering if I should fix this with those changes or leave it as-is.
